karapace/backup/v3/schema.proto

```diff
@@ -0,0 +1,147 @@
+// 2. The second byte is the backup format version, an unsigned 8-bit integer.
+//    This is always `\x03` for the backup format 3.[^1][^2]
+// 3. The next 4 bytes contain the byte count of the `Metadata` message that
+//    follows. The length is an unsigned 32-bit little-endian integer.[^3]
+// 4. The `Metadata` message.
+// 5. Zero or more records follow:
+//    1. 4 bytes containing the unsigned 32-bit little-endian integer byte count
+//       of the `Envelope` message that follows.[^3]
+//    2. The `Envelope` message.
+//
+// [^1]: The combination of these is referred to as the “magic marker version”.
+// [^2]: Endianness is irrelevant for a single byte integer.
+// [^3]: We are not using _varint_ encoding for the message framing, because it
+//   would require additional bound checks and logic to ensure that it actually
+//   is only a 32-bit value. Using fixed 4 bytes means that we can just read.
+//   Compression of the backups is used to reduce size, and we are not worried
+//   about the additional bytes required for encoding the framing bytes.
+syntax = "proto3";
+
+package karapace.backup.v3;
+
+// The metadata directly follows the magic marker version bytes. It contains
+// additional information regarding the backup.
+message Metadata {
+  // The version of the backup, this **MUST** be the same as the version that
+  // follows the magic marker byte. This is included here so that it can be
+  // passed around as part of the metadata to other parts of the application.
+  // The range of the allowed values for this is 1–255, to comply with the
+  // possible values that can be encoded in the “magic marker version” bytes.
+  // Restoration fails if the version found in the “magic marker version” and
+  // this field differ.
+  uint32 version = 1;
+
+  // The name of the tool that created the backup. The name is intended for
+  // humans, as such it **MUST** consist of printable NFC UTF-8 codepoints only.
+  // The length **MUST NOT** exceed 64 bytes.
+  //
+  // Always `Karapace` for backups created by Karapace.
+  optional string tool_name = 2;
+
+  // The version of the tool that created the backup. The version is intended
+  // for humans, as such it **MUST** consist of printable NFC UTF-8 codepoints
+  // only. The length **MUST NOT** exceed 64 bytes. The version **SHOULD** be
+  // [PEP440](https://peps.python.org/pep-0440/) or [SemVer](https://semver.org/)
+  // compatible.
+  optional string tool_version = 3;
+
+  // The UNIX time in seconds the backup creation started.
+  optional int64 create_time = 4;
+
+  // The estimated record count included in this backup. This is just an
+  // estimate, because it is not possible to determine the actual record count
+  // upfront when consuming from a topic based just on the start and end
+  // offsets. The estimation is still useful for getting an idea how many
+  // records there are, e.g. to create a progress bar.
+  optional uint64 estimated_record_count = 5;
```

Comment:
Since we're individually serializing each record, why is it not possible to count them?